feat: implementation of lua/c++ event callback system

dudantas opened the pull request:
# Description

We've implemented a callback system that allows interaction between C++ code and Lua scripts. Functions defined in C++ can be sent to a Lua script, which can manipulate them and return them to the C++ code. This allows for greater flexibility and interactivity between the two code layers, allowing functions to be manipulated at runtime by the Lua script.

Key changes include:

• Addition of the EventCallback class encapsulating callback functionality.
• Implementation of callback functions in the CallbackFunctionsBinder class.
• Registration of callback functions to the Lua interface.

These changes should be carefully reviewed to ensure the interaction between C++ and Lua is functioning as expected and that there are no memory leaks or other potential issues.

### Fixes #1228 

## Type of change
  - [x] New feature (non-breaking change which adds functionality)

## Checklist

  - [x] My code follows the style guidelines of this project
  - [x] I have performed a self-review of my own code
  - [x] I checked the PR checks reports
  - [x] I have commented my code, particularly in hard-to-understand areas
  - [x] I have made corresponding changes to the documentation
  - [x] My changes generate no new warnings

dudantas committed:
feat: implementation of lua/c++ event callback system

We've implemented a callback system that allows interaction between C++ code and Lua scripts. Functions defined in C++ can be sent to a Lua script, which can manipulate them and return them to the C++ code. This allows for greater flexibility and interactivity between the two code layers, allowing functions to be manipulated at runtime by the Lua script.

Key changes include:

• Addition of the EventCallback class encapsulating callback functionality.
• Implementation of callback functions in the CallbackFunctionsBinder class.
• Registration of callback functions to the Lua interface.

These changes should be carefully reviewed to ensure the interaction between C++ and Lua is functioning as expected and that there are no memory leaks or other potential issues.

github-actions[bot] committed:
feat: reorganize code structure and fix solution build

moviebr commented on data/libs/functions/revscriptsys.lua:
Existiria diferença de performance ao utilizar switch case ao invés de if/elseif?

beats-dh replied:
> 

A diferença de desempenho entre `switch` e `if/else if` depende de vários fatores, como o número de condições, a ordem das condições, o compilador, as otimizações do compilador, etc. No entanto, em muitos casos, a diferença de desempenho será negligenciável.

A escolha entre `switch` e `if/else if` é muitas vezes mais uma questão de legibilidade e preferência de estilo do que de desempenho. `Switch` pode ser mais limpo e fácil de ler quando você está comparando uma única variável a vários valores constantes. `If/else if`, por outro lado, é mais flexível porque pode lidar com condições mais complexas.

Lembre-se que a otimização prematura é a raiz de todo o mal (uma famosa citação de Donald Knuth). Em muitos casos, a clareza do código e a facilidade de manutenção são mais importantes do que pequenas otimizações de desempenho.

A instrução `switch` em C++ é normalmente implementada como uma "jump table". Uma "jump table" é basicamente um array de ponteiros de função, onde o valor a ser comparado é usado como índice na tabela. A execução então "salta" para a função apropriada. Isso é geralmente muito rápido, pois a busca na tabela é O(1) - constante, independentemente do número de casos.

Por outro lado, uma sequência de instruções `if/else if` é mais como uma busca linear - cada condição é avaliada por sua vez até que uma seja verdadeira, ou se esgotem as condições. Isso é O(n) - linear, com o tempo aumentando com o número de condições.

No entanto, os compiladores modernos são muito inteligentes e podem otimizar muitos usos de `if/else if` para serem tão rápidos quanto um `switch`, ou até mais rápidos em alguns casos. Por exemplo, se as condições em um `if/else if` estão em uma ordem específica que é favorável à previsão de ramificação do processador, pode ser mais rápido que um `switch`. Além disso, `if/else if` pode ser otimizado para uma busca binária se as condições estiverem em uma ordem específica, o que é O(log n) - logarítmico.

É importante também notar que `switch` tem limitações. Só pode ser usado com tipos integrais e enumerações, e só pode comparar igualdade - não pode fazer comparações complexas como maior que, menor que, ou combinar múltiplas condições.

Então, em suma, enquanto `switch` pode ser teoricamente mais rápido que `if/else if` em alguns casos devido à sua implementação como uma tabela de salto, na prática a diferença é muitas vezes negligenciável devido às otimizações do compilador. A escolha entre os dois deve ser baseada mais na legibilidade e adequação ao problema em questão do que no desempenho.

A linguagem de programação Lua, ao contrário de C++, não possui uma instrução `switch` nativa. Normalmente, as decisões condicionais são feitas usando `if`, `elseif` e `else`.

No entanto, é possível implementar uma funcionalidade semelhante à `switch` usando tabelas em Lua. Por exemplo:

```lua
action = {
  ["case1"] = function() print("case 1") end,
  ["case2"] = function() print("case 2") end,
  ["default"] = function() print("default case") end,
}

command = "case1"
(action[command] or action["default"])()
```

Isso, em certo sentido, é análogo a uma tabela de saltos, mas como Lua é uma linguagem interpretada e de tipagem dinâmica, a velocidade de execução pode variar dependendo de vários fatores, como a implementação do interpretador Lua.

A performance de uma estrutura `if-else` em Lua depende do número de condições e da ordem delas, semelhante ao que ocorre em C++. Ainda assim, a diferença de desempenho, na maioria dos casos, será pequena e a escolha deve ser feita baseada na legibilidade e na manutenção do código, ao invés do desempenho.

moviebr replied:
Perfeita explicação. Agradeço o comentário!

dudantas replied:
> Existiria diferença de performance ao utilizar switch case ao invés de if/elseif?

Não tem switch e case no lua...
Existe for que até pode simplificar, mas não vejo necessidade aqui, o lua não melhora muito a legibilidade. Veja o que fiz na função cpp que faz parse nos type, usei magic enum com um for, o que diminui de vários case ou vários elseif para poucas linhas. Mas, como falei, lua não é uma linguagem feita para otimização, infelizmente. 

#Edit
Até que consegui simplificar bastante lá, não foi usand o for, mas ficou bem melhor do que usando if e else if.

dudantas committed:
improve: better logic of functions to avoid if/elseif

dudantas committed:
feat: move onLook to revscript

moviebr reviewed (approved):
Muito bom!

dudantas committed:
Merge branch 'main' into feat-lua-event-callback-system

dudantas committed:
improve: unified callback execution function with perfect forwarding

This commit introduces a significant refactor in the way callbacks are executed. It unifies the two existing callback execution functions into a single templated function that uses perfect forwarding.

The new `executeCallback` function accepts any number of arguments of any type, and can handle both `void` and `bool` return types. It uses the C++ perfect forwarding mechanism to preserve the lvalue-ness or rvalue-ness of its arguments when they're passed to the callback function.

This change simplifies the code and reduces redundancy, while preserving the flexibility and robustness of the callback execution mechanism. It also ensures that all callbacks are executed even if some of them return `false`.

Furthermore, this commit improves the readability and maintainability of the code by removing the need for long chains of `if-else` statements in the callback registration function. Instead, a lookup table is used to map callback types to their respective functions.

github-actions[bot] committed:
Code format - (Clang-format)

dudantas committed:
fix: add clear all callbacks to the reload scripts

dudantas committed:
feat: extract some events from events.xml for example

dudantas committed:
Merge branch 'feat-lua-event-callback-system' of https://github.com/dudantas/canary into feat-lua-event-callback-system

dudantas committed:
Merge remote-tracking branch 'upstream/main' into feat-lua-event-callback-system

dudantas committed:
fix: onlook indent

dudantas committed:
fix: sonar check errors

dudantas committed:
fix: migrate setType to if

dudantas committed:
fix: move clear callbacks to the correct function

dudantas committed:
fix: add include and remove spdlog macro

dudantas committed:
feat: add readme.md explanation

dudantas committed:
Merge branch 'main' into feat-lua-event-callback-system

dudantas committed:
improve: add functions and class documentation

github-actions[bot] committed:
Code format - (Clang-format)
